src/qutip_qip/decompose/_utility.py

```diff
@@ -29,3 +32,469 @@ def check_gate(gate, num_qubits):
+    """ Finds the sequence mapping from state 1 to state 2.
+    State 1 and 2 define the basis vectors of non-trivial values in a two-level
+    unitary matrix. Here, the inputs are their respective indices in a quantum
+    gate's array.
+    This function finds the number of steps between both states when only 1 bit
+    can be changed at each step. In addition, a partial sequence going from
+    basis state 1 to state 2 and another partial sequence mapping back from
+    state 2 to state 1. The last gray sequence value in forward mapping is used
+    to define the two-level unitary.
+    """
+    # check array values are not the same
+    try:
+        assert (index_of_state_1 != index_of_state_2)
+    except AssertionError:
+        raise IndexError("Both indices need to be different.")
+
+    # array index for original binary sequence
+    indices_of_array = []
+    for i in range(2**num_qubits):
+        indices_of_array.append(i)
+
+    # Check both indices could be indices for num_qubits array
+    try:
+        assert all(x in indices_of_array for x in [
+                index_of_state_1, index_of_state_2])
+    except AssertionError:
+        raise IndexError(
+            "At least one of the input indices is invalid.")
+
+    # get the basis vector strings
+    binary_sequence_list = _binary_sequence(num_qubits)
+    state_1_binary = binary_sequence_list[index_of_state_1]
+    state_2_binary = binary_sequence_list[index_of_state_2]
+
+    # compare binary sequence positions to gray code sequence
+    # finds number of steps neeeded to go from one state to another
+    gray_code_sequence = _gray_code_sequence(num_qubits)
+    state_1_gray_code_index = gray_code_sequence.index(state_1_binary)
+    state_2_gray_code_index = gray_code_sequence.index(state_2_binary)
+    num_steps_gray_code = np.abs(
+        state_2_gray_code_index - state_1_gray_code_index)
+
+    if num_steps_gray_code == 1:
+        num_steps_gray_code = num_steps_gray_code
+    else:  # repeat the mapping back to initial index
+        num_steps_gray_code = 2*num_steps_gray_code - 1
+    # create a smaller gray code sequence between the two states of interest
+    if state_1_gray_code_index < state_2_gray_code_index:
+        gray_code = _gray_code_sequence(num_qubits)[
+            state_1_gray_code_index:state_2_gray_code_index+1]
+    else:  # check math for reversed order
+        # what is the difference between using gates from left to right vs
+        # right to left - with current code, the order is mixed.
+        gray_code = _gray_code_sequence(num_qubits)[
+            state_2_gray_code_index:state_1_gray_code_index+1]
+    gray_code_forward = gray_code
+    gray_code_backward = gray_code_forward[0:len(gray_code_forward)-1][::-1]
+
+    if num_steps_gray_code == 1:
+        return(gray_code_forward, num_steps_gray_code)
+    else:
+        return(gray_code_forward, gray_code_backward, num_steps_gray_code)
```

Comment:
The issue with creating dictionaries began here. The output was made dependent on if the gray code maps directly from one state to another or there's a couple of gates needed before implementing a two-level gate decomposition.

The new plan is to separately return the ctrl-target info for two-level gate itself and a list mapping to the two-level gate. In the case of there just being 1 step in the gray code, it would be easier to just check if list of gates is `None` or not.